simba/schedule.py

```diff
@@ -285,10 +301,170 @@ def assign_vehicles(self):
+                                             charge_curves, args)
+
+            # Add the expected soc at departure time, if greedy charging would be applied
+            socs = list(map(lambda x: partial_function(x), standing_vehicles))
+            # Pair with socs
+            standing_vehicles_w_soc = [(*v, k) for v, k in zip(standing_vehicles, socs)]
+
+            standing_vehicles_w_soc = sorted(standing_vehicles_w_soc, key=lambda x: x[-1])
+            consumption_soc = rot.calculate_consumption() / self.vehicle_types[vt][ct]["capacity"]
+            for vehicle_id, depot, soc in standing_vehicles_w_soc:
+                # end soc of vehicle if it services this rotation
+                end_soc = soc - consumption_soc
+
+                if end_soc > 0 or soc >= initial_soc or not station_is_electrified:
+                    # assign vehicles if end_soc is positive or
+                    # if soc has initial_soc.
+                    # in this case generating a new vehicle would not make a difference.
+                    # or station is not electrified. in this case vehicles should not strand there
+                    # even if rotations are not possible
+                    rot.vehicle_id = vehicle_id
+                    all_standing_vehicles.remove((vehicle_id, depot))
+                    vehicle_data[vehicle_id] = {"soc": end_soc,
+                                                "arrival_time": rot.arrival_time}
+                    break
+            else:
+                # no vehicle available for dispatch, generate new one
+                vehicle_type_counts[vt_ct] += 1
+                vehicle_id = f"{vt_ct}_{vehicle_type_counts[vt_ct]}"
+                rot.vehicle_id = vehicle_id
+                end_soc = initial_soc - consumption_soc
+                vehicle_data[vehicle_id] = {"soc": end_soc,
+                                            "arrival_time": rot.arrival_time}
+
+            rotations_in_progress.append(rot)
+            try:
+                next_rot = rotations[k + 1]
+                departure_time = next_rot.departure_time
+            except IndexError:
+                pass
+            else:
+                # keep list of rotations in progress sorted by soc reached at the time of the next
+                # departure.
+                rotations_in_progress = sorted(rotations_in_progress, key=lambda x: x.arrival_time)
+
+        # update vehicle ID for nice natural sorting
+        for rot in rotations:
+            # get old vehicle id (vehicle type + charging type, sequential number)
+            vt_ct, old_num = rot.vehicle_id.rsplit('_', 1)
+            # how many vehicles of this type?
+            vt_cnt = vehicle_type_counts[vt_ct]
+            # easy log10 to get max needed number of digits
+            digits = len(str(vt_cnt))
+            # assign new zero-padded ID (all of same vehicle type have same length)
+            rot.vehicle_id = f"{vt_ct}_{int(old_num):0{digits}}"
```

Comment:
suggestion: `rot.vehicle_id = f"{vt_ct}_{'0'*(digits-len(old_num))}{old_num}"`

Reply:
isnt f string formatting meant to do this?